### PR TITLE
cmd/snap, store, image: support for cohorts in "snap download"

### PR DESCRIPTION
**Pull request by chipaca**
This adds support for cohorts in 'snap download'.

### CODE REVIEW

File: store/store.go [hunk 1]
```diff
@@ -1262,22 +1262,6 @@ func (s *Store) WriteCatalogs(ctx context.Context, names io.Writer, adder SnapAd
 	return nil
 }
 
-// RefreshCandidate contains information for the store about the currently
```
Comment (chipaca) [inline store/store.go]: nuking this is unrelated, opportunistic

File: store/store.go [hunk 2]
```diff
@@ -1966,6 +1950,7 @@ type SnapAction struct {
 	SnapID       string
 	Channel      string
 	Revision     snap.Revision
+	CohortKey    string
```
Comment (pedronis) [inline store/store.go]: maybe it comes later but we need this also in context

File: image/helpers.go [hunk 3]
```diff
@@ -206,38 +208,81 @@ func NewToolingStore() (*ToolingStore, error) {
...
 			return "", nil, err
 		}
-		targetDir = pwd
+		opts.TargetDir = pwd
 	}
 
-	logger.Debugf("Going to download snap %q (%s) from channel %q to %q.", name, revision, opts.Channel, opts.TargetDir)
+	if !opts.Revision.Unset() {
+		// XXX: is this really necessary (and, if it is, shoudn't we error out instead)
+		opts.Channel = ""
```
Comment (pedronis) [inline image/helpers.go]: it's a long story, there are aspect to the handling of this in various places, first:

* snap install --channel=foo --revision=9999 in theory is ok  it gets a revision but is also sets the tracking channel, do we want to let people do that with download
* not relevant to this code here, but in general we had a bug at some point by which we ended up with snapst.TrackingChannel == ""

File: store/store.go (hunk 2 above)
Comment (chipaca) [reply store/store.go]: Yep, when I use it (from snapstate)

File: tests/main/snap-download/task.yaml [hunk 4]
```diff
@@ -33,3 +34,8 @@ execute: |
     su -l -c "SNAPPY_USE_STAGING_STORE=$SNAPPY_USE_STAGING_STORE HTTPS_PROXY=$HTTPS_PROXY snap download test-snapd-tools" test
     ls /home/test/test-snapd-tools_*.snap
     verify_asserts /home/test/test-snapd-tools_*.assert
+
+    echo "Snap download can download snaps from a cohort"
+    snap download --cohort="MSBlRmU4QlRSNUw1VjlGN3lIZU1BUHhrRXIyTmRVWE10dyAxNTU1NTE0MzA5IDg3ZmUwMjhkZDFjMTQ1MDY5N2QyZjdiMGZkMzgzODk0NjMzMmFhOTZmZmFjZmFlNmU2MGQyOTNjYzE1OTE3NWY=" test-snapd-tools
```
Comment (stolowski) [inline tests/main/snap-download/task.yaml]: Can we also test for errors if snap download is given revision and cohort at the same time? What about if a cohort key is invalid, can we have a test for respective error here?

File: cmd/snap/cmd_download.go [hunk 5]
```diff
@@ -103,6 +107,9 @@ func (x *cmdDownload) Execute(args []string) error {
 		if x.Channel != "" {
 			return fmt.Errorf(i18n.G("cannot specify both channel and revision"))
 		}
+		if x.CohortKey != "" {
```
Comment (stolowski) [inline cmd/snap/cmd_download.go]: Can we start growing some cli side unit tests for this? Just basic arg handling and errors for starters would be fine (a followup is ok)

File: tests/main/snap-download/task.yaml (hunk 4 above)
Comment (chipaca) [reply tests/main/snap-download/task.yaml]: You can't specify revision without being logged in, so I can't really check this from spread.
I'll add one for invalid cohort key.

File: cmd/snap/cmd_download.go (hunk 5 above)
Comment (chipaca) [reply cmd/snap/cmd_download.go]: I'll add it now. Can't do more than check errors though.

File: tests/main/snap-download/task.yaml (hunk 4 above)
Comment (chipaca) [reply tests/main/snap-download/task.yaml]: actually i can't add one for that yet because the fix for it hasn't yet been deployed

Comment (chipaca) [reply tests/main/snap-download/task.yaml]: fix got deployed, test added

File: cmd/snap/cmd_download.go (hunk 5 above)
Comment (chipaca) [reply cmd/snap/cmd_download.go]: added

File: tests/main/snap-download/task.yaml (hunk 4 above)
Comment (stolowski) [reply tests/main/snap-download/task.yaml]: Thank you!

File: cmd/snap/cmd_download.go (hunk 5 above)
Comment (stolowski) [reply cmd/snap/cmd_download.go]: Thanks!

File: cmd/snap/cmd_download_test.go [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+package main_test
+
+import (
+	"gopkg.in/check.v1"
+
+	snap "github.com/snapcore/snapd/cmd/snap"
+)
+
+// these only cover errors that happen before hitting the network,
```
Comment (stolowski) [inline cmd/snap/cmd_download_test.go]: ❤️

File: tests/main/snap-download/task.yaml [hunk 7]
```diff
@@ -33,3 +34,14 @@ execute: |
...
+    ls test-snapd-tools_*.snap
+    verify_asserts test-snapd-tools_*.assert
+    rm -v test-snapd-tools*
+
+    echo "Can't ask for invalid cohort"
+    ! snap download --cohort="what" test-snapd-tools 2>out
+    MATCH 'cannot download snap.*: Invalid cohort key' < out
```
Comment (stolowski) [inline tests/main/snap-download/task.yaml]: ❤️
